src/components/container.js

```diff
@@ -179,13 +186,167 @@ export default class Container extends Component {
+    (Store) => {
+      let containedStore = containedStores.get(Store);
+      // first time it gets called we add store to contained map bound
+      // so we can provide props to actions (only triggered by children)
+      if (!containedStore) {
+        const isExisting = registry.hasStore(Store, scope);
+        const { storeState } = registry.getStore(Store, scope);
+        const getProps = () => containerProps.current;
+        const actions = bindActions(Store.actions, storeState, getProps);
+        const handlers = bindActions(
+          { ...Store.handlers, ...override?.handlers },
+          storeState,
+          getProps,
+          actions
+        );
+        containedStore = {
+          storeState,
+          actions,
+          handlers,
+          unsubscribe: storeState.subscribe(() => handlers.onUpdate?.()),
+        };
+        containedStores.set(Store, containedStore);
+        // signal store is contained and ready now, so by the time
+        // consumers subscribe we already have updated the store (if needed)
+        if (!isExisting) handlers.onInit?.();
+      }
+      return containedStore;
+    },
+    [containedStores, registry, scope, override]
+  );
+  return [containedStores, getContainedStore];
+}
+
+function useApi(check, getContainedStore, { globalRegistry, getStore }) {
+  const getStoreRef = useRef();
+  getStoreRef.current = (Store) =>
+    check(Store) ? getContainedStore(Store) : getStore(Store);
+
+  // This api is "frozen", as changing it will trigger re-render across all consumers
+  // so we link getStore dynamically and manually call notify() on scope change
+  return useMemo(
+    () => ({ globalRegistry, getStore: (s) => getStoreRef.current(s) }),
+    [globalRegistry]
+  );
+}
+
+function createFunctionContainer({ displayName, override } = {}) {
+  const check = (store) =>
+    override
+      ? store === override.Store
+      : store.containedBy === FunctionContainer;
+
+  function FunctionContainer({ children, scope, isGlobal, ...restProps }) {
+    const ctx = useContext(Context);
+    const registry = useRegistry(scope, isGlobal, ctx);
+    const [containedStores, getContainedStore] = useContainedStore(
+      scope,
+      registry,
+      restProps,
+      override
+    );
+    const api = useApi(check, getContainedStore, ctx);
+
+    // This listens for custom props change, and so we trigger container update actions
+    // before the re-render gets to consumers, hence why memo and not effect
+    useMemo(() => {
+      containedStores.forEach(({ handlers }) => {
+        handlers.onContainerUpdate?.();
+      });
+      // Deps are dynamic because we want to notify on any custom prop change
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, Object.values(restProps).concat(containedStores));
+
+    // This listens for scope change or component unmount, to notify all consumers
+    // so all work is done on cleanup
+    useEffect(() => {
```

Comment:
Creating the store with the first call to a hook and then later on clearing it with an effect is a little weird but I don't have any better alternative as we manipulate a global state before render but then want to do cleanup after.

Reply:
Yep, we need some weird cycle to avoid context value changes and so re-render everywhere. All because React provides no sane way to opt-out...